### ISSUE TITLE
Context of input data when in edit form

### ISSUE THREAD

**Issue by ysrinu** [susanBuck/dwa15-fall2017#165]
https://github.com/susanBuck/dwa15-fall2017/blob/master/03_Laravel/99_Foobooks_Lecture_12.md#edit-a-book
recommends copying 'create' form to 'edit' form

However, the input form variables in the create form are part of $request, while in the edit mode they should be reading values from the $book. Is my assumption correct?

In my create blade form, I have
```php
<form method='POST' action='/asset'>
    {{ csrf_field() }}
    <input type='text' name='owner' id='owner' value='{{ old('owner') }}'> </br>
    <input type='submit' value='Add asset'>
</form>
```
On submit, the input value is picked in my controller store function as
`$request->input('owner');`

My asset controller edit function:
```php
public function edit($id)
{
    $result = Asset::find($id);

    if (!$result) {
        return redirect('/asset')->with('alert', 'Asset not found.');
    }

    return view('asset.edit')->with(['result' => $result]);
}
```

Now, when use the edit blade form (by cloning create blade form and editing it for 'put'):
```php
<form method='POST' action='/asset/{{ $result->id }}'>
    {{ method_field('put') }}
    {{ csrf_field() }}
    <input type='text' name='owner' id='owner' value='{{ old('owner') }}'> </br>
    <input type='submit' value='Save asset'>
</form>
```
The rendered edit form doesn't fill any values for the input 'owner' textbox even though the array $result has it (passed to it by the edit function).
This is because in the create form, there's no context for the 'owner'; it's captured in $request on submission.
However, in the edit form, the 'owner' value is in $result array.

How can we use the same (similar) form for edit that we used for create, but tell the edit form to read variables from a given array ($result here)?

**Comment by davida26** [comment]
@ysrinu If I understand correctly, you want to retrieve the data from the existing record and show it in your edit view form field. Here is what I did on my form:

```php
value="{{ $client->first_name or old('first_name') }}"
```

I also decided that creating two forms was not very DRY. So I create a form view which contains my form fields only. I then used the above to modify the value attribute. Finally, I modified my edit/create views like so:

## Edit View
```html
	<form method="POST" action="/clients/{{ $client->id }}">
        {{ method_field('put') }}
	@include('clients.form')
	</form>
```
## Create View
```html
	<form method="POST" action="/clients">
	@include('clients.form')
</div>
```

In the edit view, I used [method_field spoofing](https://laravel.com/docs/5.5/routing#form-method-spoofing) to handle the PUT method.  Of course this depends how you have your routes setup.

I hope this helps you. Made my life easier not repeating the same code :)

**Comment by ysrinu** [comment]
Thanks @davida26 
I considered using the code:
`value="{{ old('owner', $result->owner) }}`"

but it can throw error if both $owner and $result->owner doen't exist (in which case I have to use terinary operator on $result->owner).

From your example, it sounds like it might be better to use fetched value first and then resort to 'old' value if not found (mine above is in the reverse order)

Your suggestion of using clients.form is indeed very useful.

**Comment by ysrinu** [comment]
I noticed that if I use
value='{{ old('owner') or $result->owner }}'

the `old` function always returns `1` (I checked old on several variables)

If I use:
value='{{ $result->owner or old('owner')  }}'

Let's say the $result->owner is 'John' and I edit in the form to 'J' and it failed on update validation (min length should be 3, assuming the validation rules for update is different than create),
then the form shows the value 'John' and not 'J'. 
In other words the old function is kinda useless and pointless here

The other bad side effect is that on validation fail, the error displayed says the owner should be 3 characters long mini but shows the input text form value as 'John' (and not 'J' that was being submitted). This can be confusing for the end user

**Comment by davida26** [comment]
@ysrinu good catch! Yes, I agree it can be confusing for the user. This did the trick for me now on the edit view with validation:

```php
value="{{ old('company', $client->company) }}"
```
Basically, if old value exists it will show that otherwise it will default to what is stored in the DB. Handles the situation above.

In your case, try:

```php
value='{{ old('owner', $result->owner) }}'
```
and in your controller make sure to pass the $result to the view.

**Comment by susanBuck** [comment]
The above-mentioned notes ([Foobooks lecture 12](https://github.com/susanBuck/dwa15-fall2017/blob/master/03_Laravel/99_Foobooks_Lecture_12.md#edit-a-book)) don't paint the full picture of everything that was covered in lecture (it was mentioned at the top that it should not be considered a stand-alone document and to refer to the video and Foobooks code source).

In the [create](https://github.com/susanBuck/foobooks/blob/master/resources/views/book/create.blade.php) view, our inputs looked like this:
```php
<input type='text' name='title' id='title' value='{{ old('title') }}'>
```

And in the [edit](https://github.com/susanBuck/foobooks/blob/master/resources/views/book/edit.blade.php) view, the inputs ended up looking like this:
```php
<input type='text' name='title' id='title' value='{{ old('title', $book->title) }}'>
```
Reference: [Week 12, Part 3 @ 23:58](https://www.youtube.com/embed/GgiO45GiTlA?rel=0&showinfo=0&t=1434)

Regarding:
>> it can throw an error if both $owner and $result->owner don't exist (in which case I have to use ternary operator on $result->owner).

If the owner does not exist, we shouldn't be able to edit that owner. We handle this in Foobooks by redirecting the user away if they attempt to edit a book that does not exist:

```php
public function edit($id)
{
    $book = Book::find($id);

    if (!$book) {
        return redirect('/book')->with('alert', 'Book not found.');
    }

    return view('book.edit')->with(['book' => $book]);
}
```

Regarding:
>> The other bad side effect is that on validation fail, the error displayed says the owner should be 3 characters long mini but shows the input text form value as 'John' (and not 'J' that was being submitted). This can be confusing for the end user

I don't believe the current setup in Foobooks suffers from this problem. To test, I ran through the following scenario
+ I load the page to edit The Great Gatsby (<http://foobooks.dwa15.me/book/1/edit>)
+ It prefill's the title with `The Great Gatsby`
+ I change the title to `Hi`, knowing it will fail validation (because it's less than 3 chars)
+ I hit submit and am returned to the form where the error is indicated at the title field. The title field is prefilled with `Hi` again, so I can see what I had tried and fix the mistake.
+ If I hit refresh, though, it defaults back to `The Great Gatsby`, which is logical - as that's still what it is in the database.

**Comment by susanBuck** [comment]
Also: @davida26's suggestion of modularizing the form to be used for both create/edit is a good idea and, as David noted, more DRY.

If we wanted to implement this in Foobooks, the inputs should follow the style shown in the `edit` view:
```php
<input type='text' name='title' id='title' value='{{ old('title', $book->title) }}'>
```

The `create` method would just need to pass a new "empty" $book object so as to not fail when this form view modular is used for creation (otherwise the `$book->title` part will throw an error):

```php
/**
* GET /book/create
*/
public function create()
{
    $book = new Book();
    return view('book.create')->with([
        'book' => $book
    ]);
}
```

**Comment by ysrinu** [comment]
Thanks @davida26  @susanBuck 
`value='{{ old('title', $book->title) }}'`
works for the most part.

I have the below code:
Asset model:
```
public function computer()
{
    return $this->hasOne('App\Computer');
}
```

Asset controller edit function:
```
$asset= Asset::with([
    'computer',
    'group:id',
    'location:id',
    'vendor:id',
    'warranty:id',
    'outofservicecode:id'
    ])->find($id);
```
In my view blade:
`value='{{ old('memory', $asset->computer->memory) }}'`

('memory' is a column in computer table)
When an Asset has no corresponding entry in the Computer table then `$asset->computer` doesn't exist.
When the view is the first loaded:
1. If the asset has computer, it shows the input field with `$asset->computer->memory` value
2. If the asset doesn't have computer, then the view fails with error:
`Trying to get property of non-object`

My workaround is the below code:
`value='{{ old('memory', isset($asset->computer->memory) ? $asset->computer->memory : "") }}'`

Is there more leaner/elegant solution?

fyi, isset($asset->computer->memory) will be true if both $asset->computer and $asset->computer->memory exist.
If we are only checking $asset->computer then $asset->computer->memory can still fail in view if the 'memory' property doesn't exist?

I have similar problem with dropdown menus. Will post later after I spend more time debugging and if still an issue.

Thanks,
-Sri

**Comment by ysrinu** [comment]
On continuation of my above comment, after a lot of digging I figured why the dropdown values always go to default value and not retain the 'old' selected value on validation failure.
I was using '===' operator in my expression which needs both values being compared to have the same type. My value from array is of type Integer but the  'old' value is of type String. Changing to '==' fixed the issue.

**Comment by susanBuck** [comment]
Glad you were able to spot the problem @ysrinu.

To answer your above question - I can't think of leaner solution than the one you've come up with.

Closing this thread out, but feel free to start a new thread if you still have pending questions.
